### PR TITLE
[cookies] Move `YouTubeDLCookieJar` to `cookies.py`

### PR DESCRIPTION
**Pull request by coletdjnz**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

ADD DESCRIPTION HERE
Backport from #2861 

Also added a `get_cookie_header` function to the cookiejar (also from backport)

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [X] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🍪🚚🧹

This pull request refactors the cookie management in yt-dlp by moving the `YoutubeDLCookieJar` class to a separate module `yt_dlp/cookies.py` and using it directly in the `YoutubeDL` and `InfoExtractor` classes. It also adds tests and updates imports for the new module. It maintains backward compatibility for external modules that import `YoutubeDLCookieJar` from `utils`.

> _`CookieJar` of doom, you hold the secrets of the web_
> _We import you from `cookies.py`, no more legacy code to dread_
> _You simplify our extraction, you make our requests smooth_
> _`CookieJar` of doom, you are the master of our groove_

### Walkthrough
*  Move `YoutubeDLCookieJar` class from `yt_dlp/utils/_utils.py` to `yt_dlp/cookies.py` as a separate module for handling cookie files and headers ([link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-c042473d560a04c6fd504c3cb3ae2c9a56a27e42458132a265a3ffda0d60a5c7R1100-R1235), [link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-feda8f56946dc048e754111850baaef0eec4b9f8bbc2d3f04b1a785626ea5c0eL1514-L1643))
*  Update import statements in `yt_dlp/cookies.py`, `yt_dlp/extractor/common.py`, `yt_dlp/utils/_legacy.py`, and `test/test_YoutubeDLCookieJar.py` to reflect the new location of `YoutubeDLCookieJar` class ([link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-c042473d560a04c6fd504c3cb3ae2c9a56a27e42458132a265a3ffda0d60a5c7L2-R6), [link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-c042473d560a04c6fd504c3cb3ae2c9a56a27e42458132a265a3ffda0d60a5c7L32-R42), [link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-b7f6633815316063666fc3ecebb970e68fdcb5ee01beba03d04c01830e55e8a6L3443-R3443), [link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-0a7c9732b38802b4a7b67fab5d9ae6a6ff1722586503b945230e59862ac24db0R174-R178), [link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-32282851ae53d60bbdd0367e24942300f26adb82154f5e35bea30790e1d4de44L14-R14))
*  Add a wrapper function for `YoutubeDLCookieJar` in `yt_dlp/utils/_legacy.py` to maintain backward compatibility for external modules that import it from utils ([link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-0a7c9732b38802b4a7b67fab5d9ae6a6ff1722586503b945230e59862ac24db0R174-R178))
*  Modify `_get_cookies` method in `InfoExtractor` class and `_calc_headers` method in `YoutubeDL` class to use the `cookiejar` attribute instead of the `_calc_cookies` method to get the cookies for a url and generate the Cookie header ([link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-b7f6633815316063666fc3ecebb970e68fdcb5ee01beba03d04c01830e55e8a6L3443-R3443), [link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-d3ba8be45cae8dd7889a71c3360c9e4ac1160de8a5f3443b6e4a656395267f9bL2386-R2386))
*  Remove `_calc_cookies` method from `YoutubeDL` class as it is no longer needed ([link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-d3ba8be45cae8dd7889a71c3360c9e4ac1160de8a5f3443b6e4a656395267f9bL2397-L2401))
*  Add a new test case to `TestYoutubeDLCookieJar` to check the functionality of the `get_cookie_header` method ([link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-32282851ae53d60bbdd0367e24942300f26adb82154f5e35bea30790e1d4de44L50-R56))
*  Import `urllib.request` module in `yt_dlp/cookies.py` to use the `Request` class for generating the cookie header ([link](https://github.com/yt-dlp/yt-dlp/pull/7091/files?diff=unified&w=0#diff-c042473d560a04c6fd504c3cb3ae2c9a56a27e42458132a265a3ffda0d60a5c7R16))

</details>

### CODE REVIEW

File: yt_dlp/cookies.py [hunk 1]
```diff
@@ -1091,3 +1097,139 @@ def load(self, data):
...
+            for line in f:
+                try:
+                    cf.write(prepare_line(line))
+                except http.cookiejar.LoadError as e:
+                    if f'{line.strip()} '[0] in '[{"':
+                        raise http.cookiejar.LoadError(
+                            'Cookies file must be Netscape formatted, not JSON. See  '
+                            'https://github.com/yt-dlp/yt-dlp/wiki/FAQ#how-do-i-pass-cookies-to-yt-dlp')
+                    write_string(f'WARNING: skipping cookie file entry due to {e}: {line!r}\n')
+                    continue
```
Comment (Grub4K) [inline yt_dlp/cookies.py:1213]:
```suggestion
        def filter_line(line):
            if line.startswith(self._HTTPONLY_PREFIX):
                line = line[len(self._HTTPONLY_PREFIX):]
            # comments and empty lines are fine
            if line.startswith('#') or not line.strip():
                return True
            cookie_list = line.split('\t')
            if len(cookie_list) == self._ENTRY_LEN:
                cookie = _CookieFileEntry(*cookie_list)
                if cookie.expires_at and cookie.expires_at.isdigit():
                    return True
                message = f'invalid expires at {cookie.expires_at}'
            else:
                message = f'invalid length {len(cookie_list)}'

            if f'{line.strip()} '[0] in '[{"':
                raise http.cookiejar.LoadError(
                    'Cookies file must be Netscape formatted, not JSON. See  '
                        'https://github.com/yt-dlp/yt-dlp/wiki/FAQ#how-do-i-pass-cookies-to-yt-dlp')
            write_string(f'WARNING: skipping cookie file entry due to {message}: {line!r}\n')

        cf = io.StringIO()
        with self.open(filename) as file:
            for line in filter(filter_line, file):
                cf.write(line)
```
If you dont like the change, just remove the `continue` in the `except`.

**Review by Grub4K — APPROVED** [review]
Most of these are cosmetic, feel free to ignore

**Comment by coletdjnz** [comment]
> Most of these are cosmetic, feel free to ignore

I wasn't really planning on changing the cookiejar code itself in this PR to keep it simple - maybe worth refactoring it in another PR/commit?

**Comment by Grub4K** [comment]
Sure, like I said, feel free to ignore these. I can take them into another PR at some point.

**Comment by pukkandan** [comment]
Don't think this needed to be it's own PR. But since you made it, we might as well merge. After merging the relevant open PRs', can you rebase/merge #2861 and we can discuss a more reasonable way to split it up?

(PS: To clarify, I don't mind the small PRs, just think I'm adding more work for you)

**Pull request merged**